tmt/steps/provision/testcloud.py

```diff
@@ -0,0 +1,214 @@
+  - systemctl reload sshd
+"""
+
+DEFAULT_MEMORY = 2048      # MiB
+DEFAULT_DISK_SIZE = 10     # GiB
+DEFAULT_USER = 'root'
+DEFAULT_BOOT_TIMEOUT = 60  # seconds
+
+KOJI_URL = 'https://kojipkgs.fedoraproject.org/compose'
+
+RAWHIDE_URL = f'{KOJI_URL}/rawhide/latest-Fedora-Rawhide'
+RAWHIDE_ID = f'{RAWHIDE_URL}/COMPOSE_ID'
+RAWHIDE_IMAGE_URL = f'{RAWHIDE_URL}/compose/Cloud/x86_64/images'
+    
+
+def guess_image_url(name):
+    """ Guess image url for given name """
+
+    def get_compose_id(compose_id_url):
+        response = requests.get(f'{compose_id_url}')
+
+        if not response:
+            raise GeneralError(
+                f'Failed to find compose ID for '
+                f"'{name}' at '{compose_id_url}'")
+
+        return response.text
+
+    # map fedora, rawhide or fedora-rawhide to latest rawhide image
+    if re.match(r'^(fedora|fedora-rawhide|rawhide)$', name, re.IGNORECASE):
+        compose_id = get_compose_id(RAWHIDE_ID)
+        compose_name = compose_id.replace(
+            'Fedora-Rawhide', 'Fedora-Cloud-Base-Rawhide')
+        return f'{RAWHIDE_IMAGE_URL}/{compose_name}.x86_64.qcow2'
+
+    raise GeneralError("Could not map '{name}' to compose")
+
+
+class ProvisionTestcloud(ProvisionBase):
+    """ Testcloud Provisioner """
+    def __init__(self, data, step):
+        super(ProvisionTestcloud, self).__init__(data, step)
+
+        self._prepare_map = {
+            'ansible': self._prepare_ansible,
+            'shell': self._prepare_shell,
+        }
+
+        # Get image from provision options
+        if not self.option('image'):
+            raise GeneralError('No image specified')
+
+        # Initialize testcloud image 
+        self.image = None
+
+        # Testcloud instance and ip
+        self.instance = None
+        self.ip = None
+
+        # Default user
+        self.user = self.option('user') or DEFAULT_USER
+
+        # Create ssh key
+        self.ssh_key = os.path.join(self.provision_dir, 'id_rsa')
+        self.ssh_pubkey = os.path.join(self.provision_dir, 'id_rsa.pub')
+
+    def option(self, key):
+        """ Return option specified on command line """
+        # Consider command line as priority
+        if self.opt(key):
+            return self.opt(key)
+
+        return self.data.get(key, None)
+
+    def go(self):
+        super(ProvisionTestcloud, self).go()
+
+        # If image does not start with http/https/file, consider it a mapping
+        # value and try to guess the URL
+        image_url = self.option('image')
+
+        if not re.match(r'^(?:https?|file)://.*', image_url):
+            image_url = guess_image_url(image_url)
+
+        # Initialize testcloud image 
+        self.image = testcloud.image.Image(image_url)
+
+        # Show which image we are using
+        self.info('image', f'{self.image.name}', 'green')
+ 
+        status = f'{self.image.name}'
+        if not os.path.exists(self.image.local_path):
+            self.info('status', 'downloading', 'green')
```

Comment:
Download failed when running on my laptop:

    Traceback (most recent call last):
    File "/home/psss/.virtualenvs/tmt/lib/python3.7/site-packages/testcloud/image.py", line 126, in _download_remote_image
        with open(local_path + ".part", 'wb') as f:
    PermissionError: [Errno 13] Permission denied: '/var/lib/testcloud/backingstores/Fedora-Cloud-Base-Rawhide-20200228.n.0.x86_64.qcow2.part'

Shall we use a different location for storing the images?

Reply:
no, newgrp testcloud or login again .....

Reply:
should we document again?

Reply:
ah right, but if you just install testcloud from pip, I guess you need to create the location by yourself. meh. maybe we should jsut require users to install testcloud and make sure they login again (or use `newgrp testcloud`) ?

Reply:
I want to use these shared locations, as they provide caching of downloaded images ...

Reply:
But by default user is not added to the `testcloud` group. Then we should document this as well.

Reply:
hmm, I thought it is ... will document it

Reply:
fixed, testcloud user now shold not be needed. Unfortunately I had to add the domain XML template to the code, as it is added by testcloud package. Seems now we really need to install only the python package ...